tensorflow/core/lib/gif/gif_io.cc

```diff
@@ -89,23 +89,52 @@ uint8* Decode(const void* srcdata, int datasize,
   uint8* const dstdata = allocate_output(num_frames, width, height, channel);
   if (!dstdata) return nullptr;
   for (int k = 0; k < num_frames; k++) {
+    uint8* this_dst = dstdata + k * width * channel * height;
+
     SavedImage* this_image = &gif_file->SavedImages[k];
     GifImageDesc* img_desc = &this_image->ImageDesc;
+
+    int imgLeft = img_desc->Left;
+    int imgTop = img_desc->Top;
+    int imgRight = img_desc->Left + img_desc->Width;
+    int imgBottom = img_desc->Top + img_desc->Height;
+
     if (img_desc->Left != 0 || img_desc->Top != 0 || img_desc->Width != width ||
         img_desc->Height != height) {
-      *error_string = strings::StrCat("can't process optimized gif");
-      return nullptr;
+      // If the first frame does not fill the entire canvas then return error.
+      if (k == 0) {
+        *error_string =
+            strings::StrCat("the first frame does not fill the canvas");
+        return nullptr;
+      }
+      // Otherwise previous frame will be reused to fill the unoccupied canvas.
+      imgLeft = std::max(imgLeft, 0);
+      imgTop = std::max(imgTop, 0);
+      imgRight = std::min(imgRight, width);
+      imgBottom = std::min(imgBottom, height);
+
+      uint8* last_dst = dstdata + (k - 1) * width * channel * height;
```

Comment:
This is an 100% optional suggestion—probably not worth doing—of which I have little understanding. But since matrix math is happening here, and "optimized" is in the PR title, just want to point out `Tensor` (based on Eigen) and better yet MKL e.g. [mkl_matmul_op.cc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/mkl_matmul_op.cc) (which unlike Eigen doesn't appear to be available in TF builds 100% of thte time) just in case you weren't already aware.

Reply:
Further note: We have some code in tensorflow/contrib/tensorboard/db/ that might duplicate this code, except probably not? since I think that only applies to encode and not decode.

Reply:
I'm not sure what TF builds @jart is referring to. I have a hard time imagining a TF build without Eigen, since many of the core classes like ```Tensor``` are thin wrappers over Eigen classes.

In any case, I don't see any matrix math with mentioning here.